lib/mirage.ml

```diff
@@ -870,19 +960,27 @@ let stackv4_socket_conf ?(group="") interfaces = impl @@ object
 let socket_stackv4 ?group ipv4s =
   stackv4_socket_conf ?group (Key.V4.interfaces ?group ipv4s)
 
+let direct_autochooser
+    ?group ?config
+    ?(dhcp_key = Key.value @@ Key.dhcp ?group ())
+    tap =
+  if_impl Key.(pure ((=) true) $ dhcp_key)
+    (dhcp_ipv4_stack ?group tap)
+    (static_ipv4_stack ?config ?group tap)
+
 (** Generic stack *)
 
 let generic_stackv4
     ?group ?config
-    ?(dhcp_key = Key.value @@ Key.dhcp ?group ())
+    ?dhcp_key
     ?(net_key = Key.value @@ Key.net ?group ())
     (tap : network impl) : stackv4 impl =
-  if_impl
-    Key.(pure ((=) `Socket) $ net_key)
-    (socket_stackv4 ?group [Ipaddr.V4.any])
-    (if_impl Key.(pure ((=) true) $ dhcp_key)
-      (dhcp_stack ?group default_time tap)
-      (static_ipv4_stack ?config ?group tap))
+  if_impl Key.(pure ((=) `Qubes) $ value target)
+     (qubes_ipv4_stack ?group tap)
+    (if_impl Key.(pure ((=) `Socket) $ net_key)
+      (socket_stackv4 ?group [Ipaddr.V4.any])
+      (direct_autochooser ?group ?config ?dhcp_key tap)
+    )
```

Comment:
at least I'm slightly confused by the nested conditionals, which now call out to another function (only referenced once)... could we instead use a match_impl over several keys to make the code a bit clearer, please?